### ISSUE TITLE
lvm: Error when using relative lvol size

### ISSUE THREAD

**Issue by ganto** [debops/debops#939]
I recently had an error when using a relative size in the logical volume definition. The following definition should be completely valid:
```yaml
lvm__logical_volumes:
  - lv: 'prometheus'
    vg: 'data'
    size: '100%FREE'
    mount: '/var/lib/prometheus'
    mount_opts: 'relatime,discard,nodev,nosuid,noexec'
```

However on consecutive runs, Ansible bailed out with the following error (using Ansible 2.8.3 and DebOps 1.0.6):
```
TASK [debops.lvm : Manage LVM Logical Volumes] ***********************************************************************************************************************************************************************************************
task path: /usr/lib64/python3.6/site-packages/debops/ansible/roles/debops.lvm/tasks/manage_lvm.yml:43                                                                                                                                         
failed: [prometheus.example.com] (item={'lv': 'prometheus', 'vg': 'data', 'size': '100%FREE', 'mount': '/var/lib/prometheus', 'mount_opts': 'relatime,discard,nodev,nosuid,noexec'}) => {"ansible_loop_var": "item", "changed": false, "item": {"lv": "prometheus", "mount": "/var/lib/prometheus", "mount_opts": "relatime,discard,nodev,nosuid,noexec", "size": "100%FREE", "vg": "data"}, "msg": "Sorry, no shrinking of prometheus to 0 permitted."}
```

Once I added the `shrink: no` attribute to the `lvol` task, the issue was resolved.

Now I was thinking about how to properly add this to a PR. I see two different ways:
* Make `shrink` a new item attribute of `lvm__logical_volumes` which the user can explicitly set. Default would be `no` (Ansible interestingly defaults this to `yes`). I was also thinking of setting `force: yes` when shrinking is explicitly requested.
* Or simply attach the `shrink` `lvol` attribute to the `force` `lvm__logical_volumes` attribute. So when `force` is false, `shrink` would be `false` too and vice versa.

What do you think? Or should I simply add `shrink` as a new `lvm__logical_volumes` and expect the user to dig into the [`lvol` documentation](https://docs.ansible.com/ansible/latest/modules/lvol_module.html) to find the right combination how to set these?

**Comment by drybjed** [comment]
Does `shrink` require the use of `force` as well? Then I suppose those two could be tied together (your second solution).

**Comment by ganto** [comment]
Ok, the behavior is a bit unintuitive. Here the behaviour when the requested volume size (e.g. 1GB) is smaller than the actual logical volume size (e.g. 2GB):
* `force: no`/`shrink: yes` (Ansible default) - Error: _"Sorry, no shrinking of test without force=yes"_
* `force: yes`/`shrink: yes` - Success: Logical volume was resized
* `force: yes`/`shrink: no` - Success: Logcial volume was NOT resized
* `force: no`/`shrink: no` - Success: Logical volume was NOT resized

If I run the same with a relative value of '100%FREE' as logical volume size on an existing 2GB volume:
* `force: no`/`shrink: yes` (Ansible default) - First run will resize logical volume to the entire free space of the volume group. Second run will return an error: _"Sorry, no shrinking of test without force=yes"_
* `force: yes`/`shrink: yes` - First run will resize logical volume to the entire free space of the volume group. Second run will resize the volume to the original size :exclamation::question:
* `force: yes`/`shrink: no` - First run will resize logical volume to the entire free space of the volume group. Second run won't modify the size anymore.
* `force: no`/`shrink: no` - First run will resize logical volume to the entire free space of the volume group. Second run won't modify the size anymore.

This confused me a bit, because these test were not able to reproduce the exact same error message I triggered with DebOps. However, I found this interesting resize behaviour, so I repeated the test without initial volume:
* `force: yes`/`shrink: yes` - First run will create the logical volume with size of the entire free space of the volume group. Second run will return an error: _"Sorry, no shrinking of test to 0 permitted."_

Ok, here we are... But where is this size 0 coming from?

Trying to run the simple lvm command without Ansible:
```
lvs vg_test/test
  LV   VG       Attr       LSize  Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  test vg_test  -wi-a----- <1.75t

lvresize -l100%FREE vg_test/test
  New size of 0 not permitted.
```

Seems to be an LVM thing then. 

**Conclusion**
I think it's "safe" to connect the `force` and `shrink` attributes together. I guess you usually want both of them to be `no` to make sure your definition is non-destructive. Which will help in my case.

Therefore I guess we can explain `force: yes` to be destructive when used with `state: absent` (volume will be removed) and when used with `state: present` (volume might be shrinked).

**Comment by drybjed** [comment]
@ganto Excellent analysis, thanks! Perhaps this explanation could be added in its own section in `docs/ansible/roles/debops.lvm/getting-started.rst` in case that somebody looks for why the parameters are set up like that. Especially because incorrect use could result in data loss. Want to add it in the documentation?

**Comment by ganto** [comment]
Actually, I spent some more thoughts about this strange resize behaviour of lvm and I found that my intention of using the size argument "100%FREE" cannot be fulfilled by Ansible. Of course my idea is, that this LV should always take all disk space in the volume group.

The `lvol` volume will use this properly when creating a volume. However, when the volume already exists, it will again pass this argument to the `lvresize` command (as shown above) which would then be seen as an absolute size. That's why a "100%FREE" results in 0. 

What I actually wanted the module to do is, to pass "+100%FREE" in the case the volume already exists. But this behaviour obviously cannot be achieved with my inventory. The related PR allows to run exactly my use case, so that it won't throw at me that my configuration is crap. But actually it won't behave as intended, if I resize the PV and expect the LV to grow as well.

Maybe there should be an additional variable in the Ansible `lvol` module that would allow a user to express this intention? 

I guess this definitely would be worth some documentation. I will add this in the coming days.
